content/blog/2017/2017-02-02-scm-api-2-take2.adoc

```diff
@@ -0,0 +1,116 @@
++
+*Resolution:* The Branch API plugin was enhanced to detect the case where a branch source has been changed but the change is only changing the ID.
+When such changes are identified, the downstream references of the ID are all updated which will prevent a build storm.
+
+link:https://issues.jenkins-ci.org/browse/JENKINS-41313[JENKINS-41313]::
+The BitBucket Branch Source 1.x did not store all the information about PRs that is required by the SCM API 2.0.0 model.
+This could well have resulted in subtle effects when manually triggering a rebuild of a merge PR if the PR's target branch has been modified after the PR branch was first detected by Jenkins.
+Consequently, as the information is required, BitBucket Branch Source plugin 2.0.0 populated the information with dummy values which would force the correct information to be retrieved.
+The side-effect is that all PR branches would be rebuilt.
++
+*Resolution:* The changes in SCM API 2.0.2 introduced to resolve JENKINS-41121 provided a path to resolve this issue without causing a rebuild of all PR branches.
+The BitBucket Branch Source plugin was enhanced to provide an implementation of the new SCM API extension point that connects to BitBucket and retrieves the missing information.
+
+link:https://issues.jenkins-ci.org/browse/JENKINS-41124[JENKINS-41124]::
+During initial testing of the Branch API 2.0.0 release an issue was identified with how Organization Folders handled unusual names.
+None of the existing implementations of the `SCMNavigator` API could generate such unusual names due to form validation on GitHub / BitBucket replacing unusual characters with `-` when creating a repository.
++
+It would be irresponsible to rely on external services sanitizing their input data for the correct operation of Organization Folders.
+Consequently, in Branch API 2.0.0 the names were all mangled into URL safe names, with the original URLs still resolving to the original projects so that any existing saved links would remain functional.
++
+Quite a number of people objected to the change of URL scheme.
++
+*Resolution:* There has been a convention in Jenkins that the on-disk storage structure for jobs mirrors the URL structure.
+This is only a convention and there is nothing specific in the code that mandates following the convention.
++
+The Folders Plugin was enhanced to allow for computed folders (where the item names are provided by an external source) to provide a strategy to use when generating the on-disk storage names as well as the URL component names for the folder's child items.
++
+The Branch API plugin was enhanced to use this new strategy for name mangling.
++
+The net effect of this change is that the URLs remain the same as for 1.x but the on-disk storage uses mangled names that are future proofed against any new SCMNavigator implementations where the backing service allows names that are problematic to use as filesystem directory names.
```

Comment:
I wish we had a better term than "mangled". Mangled makes me think it is broken.